doc/tutorials/space_shooter/step_4.md

```diff
@@ -0,0 +1,190 @@
+  }
+}
+```
+
+The above code should be straight forward, but lets break it down:
+
+- We add to the bullet's y axis position at a rate of -500 pixels per second. Remember going up
+in the y axis means getting closer to `0` since the top left corner of the screen is `0, 0`.
+- If the y is smaller than the negative value of the bullet's height, means that the component is
+completely off the screen and it can be removed.
+
+Right, we have a `Bullet` class ready now, so lets make start to implement the action of shooting
+now. First thing, let's create two empty methods in the `Player` class, `startShooting` and
+`stopShooting`.
+
+```dart
+class Player extends SpriteAnimationComponent
+    with HasGameReference<SpaceShooterGame> {
+
+  // Rest of implementation omitted
+
+  void startShooting() {
+    // TODO
+  }
+
+  void stopShooting() {
+    // TODO
+  }
+}
+```
+
+And let's hook into those methods from the game class, we will do that by using the `onPanStart`
+and `onPanEnd` methods from the `PanDetector` mixin that we already have been using for the ship
+movement:
+
+```dart
+class SpaceShooterGame extends FlameGame with PanDetector {
+  late Player player;
+
+  // Rest of implementation omitted
+
+  @override
+  void onPanUpdate(DragUpdateInfo info) {
+    player.move(info.delta.global);
+  }
+
+  @override
+  void onPanStart(DragStartInfo info) {
+    player.startShooting();
+  }
+
+  @override
+  void onPanEnd(DragEndInfo info) {
+    player.stopShooting();
+  }
+}
+```
+
+We now have everything set up, so let's write the shooting routine in our player class.
+
+Remember, the shooting behavior will be adding bullets through time intervals when the player is
+dragging the starship.
+
+We could very much likely implement the time interval code manually, but Flame provides an component
+out of the box for that, the `TimerComponent`, so let's take advantage of it:
+
+
+```dart
+class Player extends SpriteAnimationComponent
+    with HasGameReference<SpaceShooterGame> {
+  late final TimerComponent _bulletSpawner;
+
+  @override
+  Future<void> onLoad() async {
+    // Loading animation omitted
+
+    _bulletSpawner = TimerComponent(
+      period: .2,
+      onTick: () {
+        final bullet = Bullet(
+          position: position +
+              Vector2(
+                0,
+                -height / 2,
+              ),
+        );
+        game.add(bullet);
+      },
+      repeat: true,
+      autoStart: false,
+    );
+
+    add(_bulletSpawner);
+  }
+
+  void move(Vector2 delta) {
+    position.add(delta);
+  }
+
+  void startShooting() {
+    _bulletSpawner.timer.start();
+  }
+
+  void stopShooting() {
+    _bulletSpawner.timer.stop();
+  }
+}
+```
+
+Hopefully the code above speaks for itself, but let's look at it with more detail:
+
+- First we declared a `TimerComponent` called `_bulletSpawner` in our game class, we needed it
+to be an attribute since we will accessing it in the `startShooting` and `stopShooting` methods.
```

Comment:
```suggestion
to be an variable accessible to the whole component since we will be accessing it in the `startShooting` and `stopShooting` methods.
```